GMM-HMM - experimental fix for multi-sequence fit

fcostin opened the pull request:
GMM-HMM - experimental fix for multi-sequence fit
    
Prior implementation of GMMHMM
_accumulate_sufficient_statistics had a limitation
where it failed to accumulate information from
multiple sequences during failing.
    
This caused fitting GMMHMM to multi-sequence data
not to function correctly, as it would not consider
contributions from all sequences when updating GMM
parameters during EM iterations.
    
Improves behaviour for issue 410 on github.
    
Ref: https://github.com/hmmlearn/hmmlearn/issues/410

fcostin commented on lib/hmmlearn/tests/test_gmm_hmm_multisequence.py:
```diff
@@ -0,0 +1,270 @@
+        X = np.concatenate(sequences)
+        lengths = np.asarray([len(seq) for seq in sequences])
+        converged[i], scores[i] = attempt_fit(X, lengths)
+    assert np.all(converged) and np.allclose(scores, np.mean(scores), rtol=5e-03)
```
i have no theoretical basis for choosing an rtol of 5e-3

fcostin replied:
this is better as comment in test, not in the pr.

fcostin commented on lib/hmmlearn/tests/test_gmm_hmm_multisequence.py:
```diff
@@ -0,0 +1,270 @@
+
+def attempt_fit(X, lengths):
+    model = hmm.GMMHMM(
+        n_components=2,
+        n_mix=2,
+        n_iter=100,
+        covariance_type="diag",
```
maybe should parametrise to cover all covariance_type supported by GMMHMM

fcostin commented on lib/hmmlearn/tests/test_gmm_hmm_multisequence.py:
```diff
@@ -0,0 +1,270 @@
+from hmmlearn.base import ConvergenceMonitor
+
+
+# Example multi-sequence data, arranged as shape
+# (n_sequences, n_samples, n_features)
+#
+# This data is a reduced subset of data_training.npy
+# from issue https://github.com/hmmlearn/hmmlearn/issues/410
+# illustrating GMMHMM fit diverging during EM iterations.
+#
+# Transformations to reduce data volume:
+# - keep only first 3 of many sequences, discard rest
+# - keep only first 50 of many samples per sequence, discard rest
+# - keep only first 4 of 17 features per sample, discard rest
+EXAMPLE_SEQUENCES_ISSUE_410_PRUNED = np.array([
```
this input data is not a minimal example, but is a lot smaller than the data attached to the original issue.

i don't understand GMMHMM well enough to know how to construct synthetic input that might be appropriate for testing this

anntzer commented:
This basically looks fine to me, did you plan to make other changes? (per https://github.com/hmmlearn/hmmlearn/issues/410#issuecomment-873493952)

fcostin commented:
Thank you for the review, I don't plan to make further substantial changes

> Consider fixing style per flake8, though (no need to overdo it, but at least you can try to linewrap at 79 characters).

Will do.

> Also consider squashing the commits (unless you consider the commit history important).

Will do.

Once I have fixed the style I will compress the history so there is a single test commit followed by single patch commit that makes the test pass

fcostin commented:
@anntzer requested style changes made & commit history cleaned up.

Please take another look.


When I run the test suite locally with only the new first test commit but not the second GMMHMM multi sequence patch commit, I see:

```
FAILED lib/hmmlearn/tests/test_gmm_hmm_multisequence.py::test_gmmhmm_multi_sequence_fit_invariant_to_sequence_ordering[diag]
FAILED lib/hmmlearn/tests/test_gmm_hmm_multisequence.py::test_gmmhmm_multi_sequence_fit_invariant_to_sequence_ordering[spherical]
FAILED lib/hmmlearn/tests/test_gmm_hmm_multisequence.py::test_gmmhmm_multi_sequence_fit_invariant_to_sequence_ordering[tied]
FAILED lib/hmmlearn/tests/test_gmm_hmm_multisequence.py::test_gmmhmm_multi_sequence_fit_invariant_to_sequence_ordering[full]
```

i.e. the new test is able to detect the problem for each of the four covariance types, and then passes after the patch is applied.

anntzer commented on lib/hmmlearn/tests/test_gmm_hmm_multisequence.py:
```diff
@@ -0,0 +1,290 @@
+            raise FailedToConverge("logprob must not be nan")
+
+        if self.history[-1] < self.history[-2] - self.tol:
+            raise FailedToConverge("logprob must be non-decreasing")
```
I would just return False in these two cases for the purpose of making test implementation simpler (or you can use `assert`s and rely on pytest converting assert failures to test failures).

fcostin replied:
When the monitor `converged` property returns `False` the calling `fit` method of `_BaseHMM` interprets that as "we haven't converged yet, let's keep iterating" until max iters are exhausted. The monitor doesn't seem to have a defined way to signal to `_BaseHMM` that something has gone badly wrong and the fit method abort. 

If we sense nans or decreasing logprod during iterations it seems more helpful (to maintainer trying to understand cause of test failure from symptoms) to immediately abort at first sign something is wrong instead of keeping on running.  So of the two alternatives you propose, using assertions seems superior to me than returning False, as that will abort the fit.

fcostin replied:
switching to use asserts here produces arguably superior logs when the test fails: instead of "assert all [ false false false false ]" it fails quicker and pytest dumps the entire strict monitor state that shows the logprob history values decreasing.

i've removed the exception handling. remains of `attempt_fit` method were trivial, inlined that into sole call site in the test body.

anntzer commented on lib/hmmlearn/tests/test_gmm_hmm_multisequence.py:
```diff
@@ -0,0 +1,290 @@
+
+    permutations = make_permutations(sequence_data)
+    converged = np.zeros(shape=(len(permutations),), dtype=bool)
+    scores = np.zeros(shape=(len(permutations),), dtype=np.float64)
```
converged and scores can just be lists that you append to.

anntzer commented:
Just a couple of nitpicks left.

fcostin commented:
@anntzer - thank you for the prompt review pass, i have addressed your comments, the test is a little leaner and meaner now.

if there are any remaining nits in the code that you are willing to fix up yourself, you are not stepping on my toes if you prefer to just fix them if it is a more expedient way to close off the pr

anntzer commented:
Thanks, this looks great to me!
